docs: iteration 3 (#492)
iterative/terraform-provider-iterative: README.md

![TPI](https://static.iterative.ai/img/tpi/banner.svg)

# Terraform Provider Iterative (TPI)

[![docs](https://img.shields.io/badge/-docs-5c4ee5?logo=terraform)](https://registry.terraform.io/providers/iterative/iterative/latest/docs)
[![tests](https://img.shields.io/github/workflow/status/iterative/terraform-provider-iterative/Test?label=tests&logo=GitHub)](https://github.com/iterative/terraform-provider-iterative/actions/workflows/test.yml)
[![Apache-2.0][licence-badge]][licence-file]

TPI is a [Terraform](https://terraform.io) plugin built with machine learning in mind. This CLI tool offers full lifecycle management of computing resources (including GPUs and respawning spot instances) from several cloud vendors (AWS, Azure, GCP, K8s)... without needing to be a cloud expert.

- **Lower cost with spot recovery**: transparent auto-recovery from interrupted low-cost spot/preemptible instances
- **No cloud vendor lock-in**: switch between clouds with just one line thanks to unified abstraction

> - **No waste**: auto-cleanup unused resources (terminate compute instances upon task completion/failure & remove storage upon download of results), pay only for what you use

Reviewer: 💅🏼  missing `.` periods in these

- **Developer-first experience**: one-command data sync & code execution with no external server, making the cloud feel like a laptop

> Supported cloud vendors [include][auth]:

Reviewer: 💅🏼  I'd link the other 3 words.

| ---------------------------------------------- | ---------------------------------------- | ------------------------------------------------ | ------------------------------------- |

[aws-badge]: https://img.shields.io/badge/AWS-Amazon_Web_Services-black?colorA=white&logoColor=232F3E&logo=amazonaws
[aws]: https://registry.terraform.io/providers/iterative/iterative/latest/docs/guides/authentication#amazon-web-services
[azure-badge]: https://img.shields.io/badge/Azure-Microsoft_Azure-black?colorA=white&logoColor=0078D4&logo=microsoftazure
[azure]: https://registry.terraform.io/providers/iterative/iterative/latest/docs/guides/authentication#microsoft-azure
[gcp-badge]: https://img.shields.io/badge/GCP-Google_Cloud_Platform-black?colorA=white&logoColor=4285F4&logo=googlecloud
[gcp]: https://registry.terraform.io/providers/iterative/iterative/latest/docs/guides/authentication#google-cloud-platform
[k8s-badge]: https://img.shields.io/badge/K8s-Kubernetes-black?colorA=white&logoColor=326CE5&logo=kubernetes
[k8s]: https://registry.terraform.io/providers/iterative/iterative/latest/docs/guides/authentication#kubernetes

> [auth]: https://registry.terraform.io/providers/iterative/iterative/latest/docs/guides/authentication

Reviewer: 💅🏼 💅🏼 💅🏼  Shouldn't it be at the beginning of the link list? 😋

![](https://github.com/iterative/static/raw/main/img/tpi/high-level-light.png#gh-light-mode-only)

> ![](https://github.com/iterative/static/raw/main/img/tpi/high-level-dark.png#gh-dark-mode-only)

Reviewer: > `#gh-(light|dark)-mode-only`

Woah how does that work? Just curious

## What's Special

There are a several reasons to use TPI instead of other related solutions (custom scripts and/or cloud orchestrators):

1. **Reduced management overhead and infrastructure cost**:

>    TPI is a CLI tool, not a running service. It requires no additional orchestrating machine (control plane/head nodes) to schedule/recover/terminate instances. Instead, TPI runs (spot) instances via cloud-native scaling groups[^scalers], taking care of recovery and termination automatically on the cloud provider's side. This design reduces management overhead & infrastructure costs. You can close your laptop while cloud tasks are running -- auto-recovery happens even if you are offline.

Reviewer: When we number the list it seems like we mean that it is an exhaustive list. Is it? Otherwise maybe use bullets.

But also it somewhat overlaps with the previous bullet list. Maybe they can be combined? So that the reader can get to the Usage right after the vendors figure

2. **Unified tool for data science and software development teams**:

>    TPI provides consistent tooling for both data scientists and DevOps engineers, improving cross-team collaboration. This simplifies compute management to a single config file, and reduces time to deliver ML models into production.

Reviewer 2: > _and reduces time to deliver ML models into production_

Is this a direct effect of using TPI for development? Given that this tool is not intended for model serving, that assertion migh be slightly misleading.

TPI is a [Terraform](https://terraform.io) plugin built with machine learning in mind. Full lifecycle management of computing resources (including GPUs and respawning spot instances) from several cloud vendors (AWS, Azure, GCP, K8s)... without needing to be a cloud expert.

> - **Lower cost with spot recovery**: transparent auto-recovery from interrupted low-cost spot/preemptible instances

Author: made a lot of changes since this review, a couple outstanding things:

> We are not stoping tasks, we are destroying tasks

technically we *are* stopping tasks <- terminating instances <- destroying resources

> automatic intermediate data backup

implied by transparent spot auto-recovery, right?

> - **Lower cost with spot recovery**: transparent auto-recovery from interrupted low-cost spot/preemptible instances

Reviewer 3: > implied by transparent spot auto-recovery, right?

Yes and no, you make the users to think or guess at least. There is a phase in design (title also of a praised book) "Dont make me think"

> - **Lower cost with spot recovery**: transparent auto-recovery from interrupted low-cost spot/preemptible instances

Reviewer 3: > technically we are stopping tasks

No, technically Tasks are stopped by themselves, what I want to say with this is that the user might find the ambiguity of the stop concept in cloud. Where they can stop the machine and then at some point they can restart the machine again. Here after destroy there is no possible resume

![](https://github.com/iterative/static/raw/main/img/tpi/high-level-light.png#gh-light-mode-only)
![](https://github.com/iterative/static/raw/main/img/tpi/high-level-dark.png#gh-dark-mode-only)

## What's Special

There are a several reasons to use TPI instead of other related solutions (custom scripts and/or cloud orchestrators):

1. **Reduce management overhead and infrastructure cost**:
   TPI is a CLI tool, not a running service. It requires no additional orchestrating machine (control plane/head nodes) to schedule/recover/terminate instances. Instead, TPI runs (spot) instances via cloud-native scaling groups[^scalers], taking care of recovery and termination automatically on the cloud provider's side. This design reduces management overhead & infrastructure costs. You can close your laptop while cloud tasks are running -- auto-recovery happens even if you are offline.
2. **Unified tool for data science and software development teams**:
   TPI provides consistent tooling for both data scientists and DevOps engineers, improving cross-team collaboration. This simplifies compute management to a single config file, and reduces time to deliver ML models into production.

> 

Author: ```suggestion
3. **Reproducible, codified environments**: Store hardware requirements & pipelines in a single configuration file with the rest of your ML project code.

```

TPI is a [Terraform](https://terraform.io) plugin built with machine learning in mind. Full lifecycle management of computing resources (including GPUs and respawning spot instances) from several cloud vendors (AWS, Azure, GCP, K8s)... without needing to be a cloud expert.

> - **Easy to use**: create cloud compute (CPU, GPU, RAM) & storage resources without reading pages of documentation

Author: good question.

- What are target audiences for the docs?
- What precedence do said audiences take?

For release week, I was thinking mostly driven by views :eyes: rather than bookmarks :star:

1. newbies: no cloud account (10k+ :eyes:)
2. mlops & devops: have cloud accounts and hate setup time (1k+ :eyes:)
3. ml engineers & data scientists: a mix of (1) & (2), don't need separate category

> - **Easy to use**: create cloud compute (CPU, GPU, RAM) & storage resources without reading pages of documentation

Reviewer 4: @DavidGOrtega what would be your recommendation for the 1st one?

> - **Easy to use**: create cloud compute (CPU, GPU, RAM) & storage resources without reading pages of documentation

Reviewer 3: The first issue that I find is logically, when I read it It sounds to me that we are offering the `machine` and not the `task`. 
Aside that the real problem to me is that "easy" is nice but why should I change my current aws thingy to use this. 
To me the principal motivation to do so is to know that this `task` allows me to do heavy ops in spot instances hassle-free so I would focus on what the task.
I would not put it in the first place and more after `No cloud vendor lock-in` since I would have introduced the clouds  and I would slightly rephrase it to:

**Easy to use**: Launch a task without infrastructure knowledge or having to be a Devops expert, anyone can do it.

> - **Easy to use**: create cloud compute (CPU, GPU, RAM) & storage resources without reading pages of documentation

Author: current order:

- Easy to use - newbie friendly
- Lower cost - expert friendly
- No cloud vendor lock-in - expert friendly
- Seamless developer experience - neutral
- No waste - expert friendly

is this your proposed order @DavidGOrtega?

- Lower cost - expert friendly
- Easy to use - newbie friendly
- No cloud vendor lock-in - expert friendly
- Seamless developer experience - neutral
- No waste - expert friendly

> - **Easy to use**: create cloud compute (CPU, GPU, RAM) & storage resources without reading pages of documentation

Reviewer 2: > _expert friendly_

What somebody calls “FinOps” I guess... 😅

> - **Easy to use**: create cloud compute (CPU, GPU, RAM) & storage resources without reading pages of documentation

Author: reworked this with a feature-USP split; please re-review :pray:

> - **Low cost**: transparent auto-recovery from interrupted low-cost spot/preemptible instances

Author: ```suggestion
- **Lower cost**: transparent auto-recovery from interrupted low-cost spot/preemptible instances
```

> - **No cloud vendor lock-in**: switch between several cloud vendors with ease due to concise unified configuration

Author: ```suggestion
- **No cloud vendor lock-in**: our unified abstraction allows switching between clouds with just one line
```

> - **Seamless developer experience**: easily sync & run data & code in the cloud as easily as on a local laptop

Reviewer 3: **Open source: No proprietary locks and you are able to extend it

> - **Seamless developer experience**: easily sync & run data & code in the cloud as easily as on a local laptop

Author: - dev experience: "cloud vs laptop" seems better than "CI vs laptop"
- open source: yes, but not really a USP (emphasis on "U")? Maybe could be last on the list.

> - **Seamless developer experience**: easily sync & run data & code in the cloud as easily as on a local laptop

Author: reworded & split, worth a review :pray:

- **No waste**: auto-cleanup unused resources (terminate compute instances upon job completion/failure & remove storage upon download of results)

Supported cloud vendors [include][auth]:

- Amazon Web Services (AWS)
- Microsoft Azure
- Google Cloud Platform (GCP)
- Kubernetes (K8s)

[auth]: https://registry.terraform.io/providers/iterative/iterative/latest/docs/guides/authentication

> 

Author: TODO: insert high-level daigram here

```suggestion

![](todo.png)

```

TPI is a [Terraform](https://terraform.io) plugin built with machine learning in mind. Full lifecycle management of computing resources (including GPUs and respawning spot instances) from several cloud vendors (AWS, Azure, GCP, K8s)... without needing to be a cloud expert.

- **Lower cost with spot recovery**: transparent auto-recovery from interrupted low-cost spot/preemptible instances
- **No cloud vendor lock-in**: switch between clouds with just one line thanks to unified abstraction
- **No waste**: auto-cleanup unused resources (terminate compute instances upon job completion/failure & remove storage upon download of results)
- **Seamless developer experience**: easily sync data & run code with one command, making the cloud feel like a laptop

Supported cloud vendors [include][auth]:

> - Amazon Web Services (AWS)
> - Microsoft Azure
> - Google Cloud Platform (GCP)
> - Kubernetes (K8s)

Author: technically should fit on one line:

```suggestion
Amazon Web Services (AWS) | Microsoft Azure | Google Cloud Platform (GCP) | Kubernetes (K8s)
```

Reviewer 2: Yes, sure, so nobody notices that Microsoft Azure doesn't have an official acronym[^1]

[^1]: Although the command–line tool is `az`

  }
  script = <<-END
    #!/bin/bash

    # create output directory if needed
    mkdir -p results
    # read last result (in case of spot/preemptible instance recovery)

>     if [[ -f results/epoch.txt ]]; then EPOCH="$(cat results/epoch.txt)"; fi

Reviewer 2: ```suggestion
    if test -f results/epoch.txt; then EPOCH="$(cat results/epoch.txt)"; fi
```

POSIX–friendly?

Author: I counter with https://github.com/iterative/terraform-provider-iterative/blob/docs-iter/task/common/machine/script.go#L112

Author: 1bb47a3 be happy :)

Reviewer 2: 🐲🔥✈️

Reviewer 2: If it has to be homogeneous, I don't know if it's better to use `[[` here. I tend to use `test` with `-options` and `[[` with `==` comparison operators. 🙃

  }
  script = <<-END
    #!/bin/bash

    # create output directory if needed
    mkdir -p results
    # read last result (in case of spot/preemptible instance recovery)
    if [[ -f results/epoch.txt ]]; then EPOCH="$(cat results/epoch.txt)"; fi

    # (re)start training loop up to 42 epochs

>     for epoch in $(seq ${EPOCH:-1} 42); do

Author: yes we'll definitely add the 3 pending blogs when they're merged (https://github.com/iterative/dvc.org/pull/3399, https://github.com/iterative/dvc.org/pull/3087, https://github.com/iterative/dvc.org/pull/3246)

TF_LOG_PROVIDER=INFO terraform show
```

> ### Stop Task

Reviewer: End a Task?
End the Task?

```
TF_LOG_PROVIDER=INFO terraform destroy
```

This terminates the `machine` (if still running), downloads `output`, and removes the persistent `disk_size` storage.

## How it Works

This diagram may help to see what TPI does under-the-hood:

```mermaid
flowchart LR
subgraph tpi [what TPI manages]
direction LR
    subgraph you [what you manage]
        direction LR
        A([Personal Computer])
    end
    B[("Cloud Storage (low cost)")]
    C{{"Cloud instance scaler (zero cost)"}}
    D[["Cloud (spot) Instance"]]

>     A ---> |create cloud storage| B
>     A --> |create cloud instance scaler| C
>     A ==> |upload script & workdir| B
>     A -.-> |"offline (lunch break)"| A
>     C -.-> |"(re)provision instance"| D
>     D ==> |run script| D
>     B <-.-> |persistent workdir cache| D
>     D ==> |script end,\nshutdown instance| B
>     D -.-> |outage| C
>     B ==> |download output| A

Author: ```suggestion
    A ---> |2. create cloud storage| B
    A --> |1. create cloud instance scaler| C
    A ==> |3. upload script & workdir| B
    A -.-> |"4. offline (lunch break)"| A
    C -.-> |"5. (re)provision instance"| D
    D ==> |7. run script| D
    B <-.-> |6. persistent workdir cache| D
    D ==> |8. script end,\nshutdown instance| B
    D -.-> |outage| C
    B ==> |9. download output| A
```

Reviewer 4: Looks good.

Reviewer: Cool diagram but it's not readable without clicking the `<->` button (not necessarily obvious usability). Any chance to make it taller and less long i.e. fold the diagram somehow?

end
style you fill:#FFFFFF00,stroke:#13ADC7
style tpi fill:#FFFFFF00,stroke:#FFFFFF00,stroke-width:0px
style A fill:#13ADC7,stroke:#333333,color:#000000
style B fill:#945DD5,stroke:#333333,color:#000000
style D fill:#F46737,stroke:#333333,color:#000000
style C fill:#7B61FF,stroke:#333333,color:#000000
```

## Future Plans

TPI is a CLI tool bringing the power of bare-metal cloud to a bare-metal local laptop. We're working on more featureful and visual interfaces. We'd also like to have more native support for distributed (multi-instance) training, more data sync optimisations & options, and tighter ecosystem integration with tools such as [DVC](https://dvc.org).

## Help

> The [getting started guide](https://registry.terraform.io/providers/iterative/iterative/latest/docs/guides/getting-started) has some more information. In case of errors, extra debugging information is available using `TF_LOG_PROVIDER=DEBUG` instead of `INFO`.

Reviewer: 💅🏼 

```suggestion
The [Getting Started](https://registry.terraform.io/providers/iterative/iterative/latest/docs/guides/getting-started) guide has some more information. In case of errors, extra debugging information is available using `TF_LOG_PROVIDER=DEBUG` instead of `INFO`.
```

   ```
3. Use `source = "github.com/iterative/iterative"` in your `main.tf` to use the local repository (`source = "iterative/iterative"` will download the latest release instead), and run `terraform init --upgrade`

> This diagram may also help to see what TPI does under-the-hood:

Reviewer 3: ```suggestion
## How does it work

This diagram may also help to see what TPI does under-the-hood:
```

Reviewer 3: Prior to ## Contributing

```mermaid
flowchart LR
subgraph tpi [what TPI manages]
direction LR
    subgraph you [what you manage]
        direction LR
        A([Personal Computer])
    end
    B[("Cloud Storage (low cost)")]
    C{{"Cloud Orchestrator (zero cost)"}}
    D[["Cloud (spot) Instance"]]
    A ---> |create cloud storage| B
    A --> |create cloud orchestrator| C
    A ==> |upload script & workdir| B
    A -.-> |"offline (lunch break)"| A
    C -.-> |"(re)provision instance"| D
    D ==> |run script| D
    B <-.-> |persistent workdir cache| D
    D ==> |script end,\nshutdown instance| B
    D -.-> |outage| C
    B ==> |download output| A
end

> style you fill:#fff,stroke:#13ADC7
> style tpi fill:#fff,stroke:#fff,stroke-width:0px

Author: ```suggestion
style you fill:#FFFFFF,stroke:#13ADC7
style tpi fill:#FFFFFF,stroke:#FFFFFF,stroke-width:0px
```

> style you fill:#fff,stroke:#13ADC7
> style tpi fill:#fff,stroke:#fff,stroke-width:0px
> style A fill:#13ADC7,stroke:#333
> style B fill:#945DD5,stroke:#333
> style D fill:#F46737,stroke:#333
> style C fill:#7B61FF,stroke:#333

Author: ```suggestion
style you fill:#FFFFFF00,stroke:#13ADC7
style tpi fill:#FFFFFF00,stroke:#FFFFFF00,stroke-width:0px
style A fill:#13ADC7,stroke:#333333,color:#000000
style B fill:#945DD5,stroke:#333333,color:#000000
style D fill:#F46737,stroke:#333333,color:#000000
style C fill:#7B61FF,stroke:#333333,color:#000000
```
